Copy local intellisense xmls for assemblies with source of truth.

carlossanlop opened the pull request:
This PR explores obtaining the intellisense xml files for assemblies that have their source of truth in triple slash from the local build, and the rest of the assemblies get their intellisense xml files from the nuget package that comes from the docs repo.

Made sure the `dotnet pack` command also included the correct intellisense file in the nupkgs.

carlossanlop committed:
Copy local intellisense xmls for assemblies with source of truth.

smasher164 commented:
The CI errors seem to be happening because it's trying to copy the XMLs over on configurations besides `-allconfigurations`, which shouldn't happen. Maybe the `docsFile` target runs even when not in `allconfigurations`?

carlossanlop commented:
Our assumption was that we only want this xml copy to happen when `-allconfigurations` is executed because when we do not build `-allconfigurations`, the only folder that gets created for many assemblies is the `net7.0-windows` one, but we need the non-OS specific one for this to work. We could be wrong about our approach, though.

If we confirm we only want this copy to happen in `-allconfigurations`, then we need to find a way to detect that the user passed `-allconfigurations` and use it as a condition to run our target.

smasher164 committed:
use UseIntellisenseDocumentationFile instead

carlossanlop committed:
Create intellisense.targets, which contains the Target that defines the set of files to copy, and the Target that copies each of those files into the artifacts/bin/docs folder.
The paths of the files to copy are dynamically defined depending if the assembly IsPackable property is set or not, and if the UseIntellisenseDocumentationFile boolean is set or not.

carlossanlop committed:
Set UseIntellisenseDocumentationFile to true for the 3 assemblies that already have their source of truth in triple slash.

carlossanlop commented on Directory.Build.props:
```diff
@@ -103,7 +103,7 @@
     <MibcOptimizationDataDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'mibc'))</MibcOptimizationDataDir>
     <XmlDocDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'docs'))</XmlDocDir>
-    <XmlDocFileRoot>$([MSBuild]::NormalizeDirectory('$(NuGetPackageRoot)', 'microsoft.private.intellisense', '$(MicrosoftPrivateIntellisenseVersion)', 'IntellisenseFiles', 'net'))</XmlDocFileRoot>
+    <XmlDocFileRoot>$([MSBuild]::NormalizeDirectory('$(NuGetPackageRoot)', 'microsoft.private.intellisense', '$(MicrosoftPrivateIntellisenseVersion)', 'IntellisenseFiles'))</XmlDocFileRoot>
```
``The `net` folder only applies for in-box assemblies. OOB assemblies have their documentation in the `dotnet-plat-ext` folder.

Instead of hardcoding the last leaf of the path here, we remove it and will define it later when this value is needed (only used in one place).

carlossanlop commented on Directory.Build.targets:
```diff
@@ -12,6 +12,7 @@
   <Import Project="$(RepositoryEngineeringDir)generators.targets" />
   <Import Project="$(RepositoryEngineeringDir)python.targets" />
   <Import Project="$(RepositoryEngineeringDir)resolveContract.targets" Condition="'$(IsSourceProject)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)intellisense.targets" Condition="'$(MSBuildProjectExtension)' != '.pkgproj'" />
```
Instead of adding the intellisense Targets in packaging.targets, it feels cleaner to have them defined in a separate file. Since some of the elements defined in this new file are required by packaging.targets, the Import is added before packaging targets. The condition seems to make sense here too.

carlossanlop commented on eng/packaging.targets:
```diff
@@ -28,9 +28,7 @@
-    <!-- Search for the documentation file in the intellisense package and otherwise pick up the generated one. -->
-    <LibIntellisenseDocumentationFilePath>$(XmlDocFileRoot)1033\$(AssemblyName).xml</LibIntellisenseDocumentationFilePath>
-    <UseIntellisenseDocumentationFile Condition="'$(UseIntellisenseDocumentationFile)' == '' and Exists('$(LibIntellisenseDocumentationFilePath)')">true</UseIntellisenseDocumentationFile>
+
```
These were moved to intellisense.targets. Since it's an Import, and is added before packaging.targets, they can be accessed here too.

carlossanlop commented on eng/packaging.targets:
```diff
@@ -108,16 +106,6 @@
                                        ($(TargetFrameworks.Contains('$(NetFrameworkMinimum)')) or $(TargetFrameworks.Contains('net47')) or $(TargetFrameworks.Contains('net48')))" />
   </ItemGroup>
 
-  <!-- TODO: Remove this target when no library relies on the intellisense documentation file anymore.-->
```
Moved to intellisense.targets.

carlossanlop commented on eng/restore/docs.targets:
```diff
@@ -4,43 +4,4 @@
     <PackageDownload Include="Microsoft.Private.Intellisense" Version="[$(MicrosoftPrivateIntellisenseVersion)]" />
   </ItemGroup>
 
-  <!-- the intellisense package doesn't use nuget conventions so we need to select manually -->
```
Not needed anymore. We don't have localized folders in the nupkg anymore, and we are going to decide which xml to copy (from nupkg or from local build) inside intellisense.targets, not here.

We still need the first few lines that will download the package. Maybe we could move it somewhere else so this file gets deleted?

ViktorHofer replied:
Please move this PackageDownload item into the Build.proj file in the repo root, delete this file and remove the import to it.

carlossanlop replied:
Done.

carlossanlop commented on src/libraries/System.Formats.Cbor/src/System.Formats.Cbor.csproj:
```diff
@@ -3,6 +3,7 @@
     <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <IsPackable>true</IsPackable>
+    <UseIntellisenseDocumentationFile>true</UseIntellisenseDocumentationFile>
```
These 3 csproj files are supposed to have their documentation source of truth already in triple slash comments. But in a future PR, I will do an additional pass with api-docs-sync to ensure the docs are fully synced.

I tested by modifying a triple slash comment of a public API, and I could confirm that the final file in artifacts/bin/docs contained the change.

carlossanlop commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+
+  <PropertyGroup>
+      <!-- Search for the documentation file in the intellisense package and otherwise pick up the generated one. -->
+    <IntellisensePackageFolderName Condition="'$(IsPackable)' == '' or '$(IsPackable)' == 'false'">net</IntellisensePackageFolderName>
+    <IntellisensePackageFolderName Condition="'$(IsPackable)' == 'true'">dotnet-plat-ext</IntellisensePackageFolderName>
```
Instead of having a hardcoded`'net'` value in `Directory.Build.props`, we decide here the value we want.

There are other folders in the intellisense nupkg: `netstandard` and `windowsdesktop`. I am unsure if we need to add more conditions to also include those.

I suspect the `windowsdesktop` folder would have to be used in the winforms repo infrastructure, not here.

smasher164 replied:
Yeah this is the only part I'm unsure about. Maybe we'll need those other folders as well.

ViktorHofer replied:
You shouldn't need netstandard as it only contains the netstandard folder as it only contains the netstandard.xml which we don't expose in our targeting pack. It was only used by the netstandard repo, which is now archived.

Neither do I think that you need windowsdesktop.

But as Eric said in the stand-up on Monday, we shouldn't condition on `IsPackable` but probe in a priority order instead:
1. Probe inside the "net" folder
2. If the file doesn't exist, probe the "dotnet-plat-ext" folder.
3. If the file still doesn't exist, use the compiler generated xml file.

But we should only do that when `$(UseIntellisenseDocumentationFile)` isn't true as that probing would otherwise be a waste of time.

carlossanlop replied:
This should be addressed.

carlossanlop commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+      <!-- Search for the documentation file in the intellisense package and otherwise pick up the generated one. -->
+    <IntellisensePackageFolderName Condition="'$(IsPackable)' == '' or '$(IsPackable)' == 'false'">net</IntellisensePackageFolderName>
+    <IntellisensePackageFolderName Condition="'$(IsPackable)' == 'true'">dotnet-plat-ext</IntellisensePackageFolderName>
+    <LibIntellisenseDocumentationFilePath Condition="'$(UseIntellisenseDocumentationFile)' == 'true'">$([MSBuild]::NormalizePath('$(ArtifactsBinDir)', '$(MSBuildProjectName)', '$(LibrariesConfiguration)', '$(NetCoreAppCurrent)', '$(MSBuildProjectName).xml'))</LibIntellisenseDocumentationFilePath>
+    <LibIntellisenseDocumentationFilePath Condition="'$(UseIntellisenseDocumentationFile)' == '' or '$(UseIntellisenseDocumentationFile)' == 'false'">$([MSBuild]::NormalizePath('$(XmlDocFileRoot)', '$(IntellisensePackageFolderName)', '1033', '$(AssemblyName).xml'))</LibIntellisenseDocumentationFilePath>
```
Let me know if there's a better way to determine the variable value of these properties.

carlossanlop replied:
Eric gave me feedback on this in the comment above: https://github.com/dotnet/runtime/pull/79134/files#r1067578604

carlossanlop commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+
+  <!-- TODO: Remove this target when no library relies on the intellisense documentation file anymore.-->
+  <Target Name="ChangeDocumentationFileForPackaging"
+          AfterTargets="DocumentationProjectOutputGroup"
```
What is `DocumentationProjectOutputGroup`? This was defined here (@smasher164, @ViktorHofer) but I was unable to find any target with that name.

carlossanlop replied:
This is a super n00b question. I _thiink_ the `Output` suffix gets removed from the `ItemGroup` elements, allowing access to them with `DocumentationProjectOutputGroup`, instead of `DocumentationProjectOutputGroupOutput`. Is this correct?

ViktorHofer replied:
This is just an inconsistency in the naming of the target vs the item: https://github.com/dotnet/msbuild/blob/a2490dd3f78cce4abc8f9e6f1b5268437332818f/src/Tasks/Microsoft.Common.CurrentVersion.targets#L6244-L6246

That target and item is defined by msbuild's common targets layout.

carlossanlop commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+          AfterTargets="DocumentationProjectOutputGroup"
+          Condition="'$(UseIntellisenseDocumentationFile)' == 'true'">
+    <ItemGroup>
+      <DocumentationProjectOutputGroupOutput Remove="@(DocumentationProjectOutputGroupOutput)" />
```
@smasher164 and I were wondering what the purpose of this remove is. I suspect the purpose is to clean the item group if it has previous values, so that it gets filled out again from scratch. Correct?

ViktorHofer replied:
That item by default contains the compiler generated file. As NuGet uses that item for packaging, we clear it out and instead add the path to the intellisense file from the nuget package.

ericstj commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+          AfterTargets="CoreCompile"
+          Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'"
+          DependsOnTargets="ChangeDocumentationFileForPackaging">
+    <Copy SourceFiles="@(DocumentationProjectOutputGroupOutput)"
```
This is failing because `DocumentationProjectOutputGroupOutput` is pointing to the path in the bin folder, but that doesn't exist at the time that `CoreCompile` completes.  Instead it's only in the `obj` folder.  

I think we can run this later in the build, after the copy would normally happen.  Looks like this happens as part of `CopyFilesToOutputDirectory` : https://github.com/dotnet/msbuild/blob/a2490dd3f78cce4abc8f9e6f1b5268437332818f/src/Tasks/Microsoft.Common.CurrentVersion.targets#L4770

ericstj replied:
This is still a problem.  Fix is also suggested above.

ericstj commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+
+  <Target Name="CopyDocumentationFileToXmlDocDir"
+          AfterTargets="CoreCompile"
+          Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'"
```
I think this should only run for src projects.  I think this ends up running in `ref` projects right now and doing nothing since the `DocumentationProjectOutputGroupOutput` item is empty.  You probably want to condition this to only run for src projects since you wouldn't want two projects with the same name copying xml files to a single directory.  We're also only expecting the src projects to actually contain the docs.  `IsSourceProject` should probably be considered here.

carlossanlop replied:
Such condition would be: `'$(IsSourceProject)' == 'true'`, is that right?

Should I add the condition only to this target? Or should I add such condition to the whole new `intellisense.targets` import file?

I ask because in `Directory.Build.targets`, this import was added with a condition to skip pkgproj extension projects, but I could also add the src condition you're talking about there.

ericstj replied:
You could do that, it's more of a style question.  @ViktorHofer do you have any style preferences for understanding the imports in our repo?  Do we want conditions on imports and targets, or would you prefer to just place the conditions on targets to avoid mistakes where folks assume targets are imported?

carlossanlop replied:
Well, the style is already there, where we import intellisense.targets. I just extracted the intellisense-specific Targets from the old file and put it in a new one. My understanding is that the condition for importing both *.targets files should be the same, but I need clarification in this other comment: https://github.com/dotnet/runtime/pull/79134#discussion_r1068378602

ericstj commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+      <!-- Search for the documentation file in the intellisense package and otherwise pick up the generated one. -->
+    <IntellisensePackageFolderName Condition="'$(IsPackable)' == '' or '$(IsPackable)' == 'false'">net</IntellisensePackageFolderName>
+    <IntellisensePackageFolderName Condition="'$(IsPackable)' == 'true'">dotnet-plat-ext</IntellisensePackageFolderName>
+    <LibIntellisenseDocumentationFilePath Condition="'$(UseIntellisenseDocumentationFile)' == 'true'">$([MSBuild]::NormalizePath('$(ArtifactsBinDir)', '$(MSBuildProjectName)', '$(LibrariesConfiguration)', '$(NetCoreAppCurrent)', '$(MSBuildProjectName).xml'))</LibIntellisenseDocumentationFilePath>
```
I don't think you should be defining this property to point at the built file.  Instead just rely on the value set by `DocumentationProjectOutputGroup` in the item `DocumentationProjectOutputGroupOutput`.

Let `LibIntellisenseDocumentationFilePath` only point to the nuget-package sourced version of the xml file.

You don't need to have this property be conditioned on `UseIntellisenseDocumentationFile` - that's redundant.  The only thing that needs to check `UseIntellisenseDocumentationFile` is the target that swaps the built xml with the package one.

ViktorHofer replied:
Agreed. We don't want to hardcode the output paths anywhere as they are subject to change.

ericstj commented:
Things to test to make sure this is behaving correctly:
1. Library nuget package is honoring this flag and packing source built XML vs doc-team provided XML.  Check both cases.
2. Microsoft.NETCore.App targeting pack is honoring this flag.  Check both cases.
3. Transport packages are honoring this flag.  Check both cases.

I don't *think* any of these are broken, but I'm not sure about the last one.  It would be good to check.

ViktorHofer commented:
I just submitted https://github.com/dotnet/arcade/pull/12188 to make it possible to have a compiler error for NoWarn=CS1591. For historical reasons that error is globally suppressed in all Arcade.Sdk consuming repositories.

We should set `<SkipArcadeNoWarnCS1591>true</SkipArcadeNoWarnCS1591>` in Versions.props and then set `<NoWarn Condition="'$(UseIntellisenseDocumentationFile)' != 'true'">$(NoWarn);1591</NoWarn>` in intellisense.targets to not break existing projects that haven't yet annotated all their public API.

ViktorHofer commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+<Project>
+
+  <PropertyGroup>
+      <!-- Search for the documentation file in the intellisense package and otherwise pick up the generated one. -->
```
nit: indentation is off

```suggestion
    <!-- Search for the documentation file in the intellisense package and otherwise pick up the generated one. -->
```

ViktorHofer commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,32 @@
+
+  <Target Name="CopyDocumentationFileToXmlDocDir"
+          AfterTargets="CoreCompile"
+          Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'"
```
You only want to copy the xml file if the targeting pack needs it. That's more or less determined via the following property: `IsNetCoreAppSrc`: https://github.com/dotnet/runtime/blob/25e84d3ede510e052bf24c80bc3ce92f7c784e0f/src/libraries/Directory.Build.targets#L57

I think it's OK to require a project to target `$(NetCoreAppCurrent)` when it's inbox but we might need to correct that in the future. Honestly, I believe we will change this anyway when we switch to ProjectReferences in the sfxproj but as we aren't there yet, this should suffice.

So you want to update your condition to: `'$(IsNetCoreAppSrc)' == 'true' and '$(TargetFramework)' == '$(NetCoreAppCurrent)'`

carlossanlop commented on Directory.Build.props:
```diff
@@ -103,7 +103,7 @@
     <MibcOptimizationDataDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'mibc'))</MibcOptimizationDataDir>
     <XmlDocDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'docs'))</XmlDocDir>
-    <XmlDocFileRoot>$([MSBuild]::NormalizeDirectory('$(NuGetPackageRoot)', 'microsoft.private.intellisense', '$(MicrosoftPrivateIntellisenseVersion)', 'IntellisenseFiles', 'net'))</XmlDocFileRoot>
+    <XmlDocFileRoot>$([MSBuild]::NormalizeDirectory('$(NuGetPackageRoot)', 'microsoft.private.intellisense', '$(MicrosoftPrivateIntellisenseVersion)', 'IntellisenseFiles'))</XmlDocFileRoot>
```
Yep. I was doing that just yesterday. Realized it would be best to have that property in the other file. It seems that XmlDocDir (the one above) could also be moved to that other file. We only use it there.

ViktorHofer replied:
The other property should stay here and https://github.com/dotnet/runtime/blob/2038514d8dff40dc7add9e89389c856f73125cb2/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Ref.sfxproj#L17 should actually use it instead of hardcoding the path.

carlossanlop replied:
Ah good find. I'll leave it here and replace the usage you pointed out.

carlossanlop commented on Directory.Build.targets:
```diff
@@ -12,6 +12,7 @@
   <Import Project="$(RepositoryEngineeringDir)generators.targets" />
   <Import Project="$(RepositoryEngineeringDir)python.targets" />
   <Import Project="$(RepositoryEngineeringDir)resolveContract.targets" Condition="'$(IsSourceProject)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)intellisense.targets" Condition="'$(MSBuildProjectExtension)' != '.pkgproj'" />
```
@ViktorHofer sure, I'll add that condition. But why do we not want to preserve the one for pkgproj? That condition is used by the import in the next line (packaging.targets) which is where we originally had the intellisense-related work located.

ViktorHofer replied:
IsSourceProject already means no pkgproj: https://github.com/dotnet/runtime/blob/5fda6fd23fa1d2732ca3418d5b6829dfd1e61a19/Directory.Build.props#L350-L359

carlossanlop committed:
Delete docs.targets, move package download to Build.proj

carlossanlop committed:
Consume $(XmlDocDir) in Microsoft.NetCore.App.Ref.sfxproj

carlossanlop committed:
Change condition to import intellisense.targets: Use $(IsSourceProject) == true

carlossanlop committed:
Move XmlDocFileRoot to intellisense.targets, address suggestions for collecting xml file.

smasher164 commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,41 @@
+  </Target>
+
+  <Target Name="CopyDocumentationFileToXmlDocDir"
+          AfterTargets="CoreCompile"
```
IIRC, @ericstj was mentioning that the `AfterTargets` property here shouldn't be `CoreCompile`, since the `DocumentationProjectOutputGroupOutput` is empty.

ericstj replied:
Not necessarily because `DocumentationProjectOutputGroupOutput`.  That will get set because this depends on `ChangeDocumentationFileForPackaging` which in turn depends on `DocumentationProjectOutputGroup` which sets `DocumentationProjectOutputGroupOutput`.  The problem is that the file isn't yet in the final location after `CoreCompile`, which is why I made [this comment](https://github.com/dotnet/runtime/pull/79134/files#r1067565007).  As I mentioned there, you should make this `AfterTargets=CopyFilesToOutputDirectory`.
```suggestion
          AfterTargets="CopyFilesToOutputDirectory"
```

ericstj commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,41 @@
+  </Choose>
+
+  <!-- TODO: Remove this target when no library relies on the intellisense documentation file anymore.-->
+  <Target Name="ChangeDocumentationFileForPackaging"
```
As we just discussed, this might be too late to impact the result of a ProjectReference.

ProjectReferences (as used in the transport packages) only recieve the path to the dll in the bin folder and find the doc file relative to that.

To support  ProjectReference we may want to swap the value of `@(DocFileItem)` instead and do so right before `CopyFilesToOutputDirectory`.  This will make it so that the "shipping" xml is always next to the binary in the bin folder.  The compiler generated one (in the case it is not shipping) will remain in the obj folder.

ericstj replied:
Here's a sample that does this: https://github.com/ericstj/runtime/commit/94849babba1fdcffec9a5a53fc0792377e0ae7fb

ViktorHofer replied:
Die you verify that this works in a pack-only scenario as well? dotnet pack --no-build

carlossanlop replied:
Added some tests in a comment in the main thread.

carlossanlop commented:
> I just submitted [dotnet/arcade#12188](https://github.com/dotnet/arcade/pull/12188) to make it possible to have a compiler error for NoWarn=CS1591

Thanks, @ViktorHofer. I'll rebase the PR.

carlossanlop committed:
Expand condition of CopyDocumentationFileToXmlDocDir to also check if IsNetCoreAppSrc == true.

carlossanlop committed:
Change AfterTargets of CopyDocumentationFileToXmlDocDir from CoreCompile to CopyFilesToOutputDirectory

carlossanlop commented:
After the last pushes, I am seeing the following behaviors locally after building everything in release mode:

- Fail? - In artifacts/bin/docs, I am not seeing the xmls for `System.IO.Compression.Brotli`, `System.Formats.Cbor` and `System.Numerics.Vectors`. Coincidentally, those are the 3 csprojects that have the `<UseIntellisenseDocumentationFile>` flag set to `true`.
- Success - In `artifacts/bin/System.Formats.Cbor/Release/net8.0`, I am seeing the locally build xml placed there. I know this because I temporarily changed the summary of a public API in that assembly to something unique, and I'm seeing that text in the copied file.
- Success - In `artifacts/bin/System.IO.Compression.Brotli/Release/net8.0-windows` I am seeing the locally build xml placed there.
- Fail? - But in `artifacts/bin/System.IO.Compression.Brotli/Release/net8.0` (generic platform) I am seeing an xml with only `SR` texts.
- Fail? - After running `dotnet pack -c release` inside `src/libraries/System.Formats.Cbor`, and opening the generated nuget package `artifacts\packages\Release\Shipping\System.Formats.Cbor.8.0.0-dev.nupkg`, I do not see any xmls anywhere inside.

ViktorHofer commented on Build.proj:
```diff
@@ -10,7 +10,10 @@
 
   <!-- Upfront restore hooks -->
-  <Import Project="$(RepositoryEngineeringDir)restore\docs.targets" />
+  <ItemGroup>
+    <PackageDownload Include="Microsoft.Private.Intellisense" Version="[$(MicrosoftPrivateIntellisenseVersion)]" />
+  </ItemGroup>
```
Actually, can you please move that `PackageDownload` item into intellisense.targets? We want to inline this dependency into the project's restore phases that need it. That avoid an artificial upfront restore requirement.

carlossanlop replied:
Good point. Makes sense.

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Ref.sfxproj:
```diff
@@ -14,7 +14,7 @@
     <ItemGroup>
       <ReferencePath Include="@(LibrariesRefAssemblies)" />
-      <DocFilesToPackage Include="$(ArtifactsBinDir)/docs/%(LibrariesRefAssemblies.FileName).xml" Condition="Exists('$(ArtifactsBinDir)/docs/%(LibrariesRefAssemblies.FileName).xml')"/>
+      <DocFilesToPackage Include="$(XmlDocDir)/%(LibrariesRefAssemblies.FileName).xml" Condition="Exists('$(ArtifactsBinDir)/docs/%(LibrariesRefAssemblies.FileName).xml')"/>
```
```suggestion
      <DocFilesToPackage Include="$(XmlDocDir)%(LibrariesRefAssemblies.FileName).xml" Condition="Exists('$(XmlDocDir)%(LibrariesRefAssemblies.FileName).xml')"/>
```

ViktorHofer replied:
Unresolving as my suggested change isn't yet addressed.

carlossanlop replied:
Aah sorry, I missed the Condition's property. I only changed it in the Include. Thanks for noticing.

ViktorHofer commented on src/libraries/System.IO.Compression.Brotli/src/System.IO.Compression.Brotli.csproj:
```diff
@@ -2,6 +2,7 @@
   <PropertyGroup>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-unix;$(NetCoreAppCurrent)</TargetFrameworks>
+    <UseIntellisenseDocumentationFile>true</UseIntellisenseDocumentationFile>
```
This property's name doesn't really help you understand what it does. I think something like `UseIntellisensePackageDocumentationFile` would make more sense.

carlossanlop replied:
Should we use the reserved one instead? https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#generatedocumentationfile

ViktorHofer replied:
Probably not as we always want to generate documentation files regardless of whether they are used or not.

ericstj replied:
It's also not clear to me what you actually want he behavior to be here.  I see these projects set this to `true` and setting to `true` causes the private.intellisense package to be used -- which I thought was the default?

ViktorHofer replied:
Agreed. Let's default `UseIntellisensePackageDocumentationFile` to true in the intellisense.targets file (if not previously set) and make these project files set it to false instead.

carlossanlop replied:
Ok sure.

carlossanlop replied:
> Let's default `UseIntellisensePackageDocumentationFile` to true in the intellisense.targets file (if not previously set)

Ah, I missed the "if not previously set" part.

ViktorHofer commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,41 @@
+
+  <!-- If the project is not generating its own documentation yet, retrieve it from the dotnet-api-docs nuget package. -->
+  <Choose>
+    <When Condition="('$(UseIntellisenseDocumentationFile)' == '' or '$(UseIntellisenseDocumentationFile)' == 'false') and Exists($([MSBuild]::NormalizePath('$(XmlDocFileRoot)', 'net', '1033', '$(AssemblyName).xml')))">
```
MSBuild style nit: Properties that are used for boolean feature flags should only contain two values: empty (which means false) or 'true'.

```suggestion
    <When Condition="'$(UseIntellisenseDocumentationFile)' != 'true' and Exists($([MSBuild]::NormalizePath('$(XmlDocFileRoot)', 'net', '1033', '$(AssemblyName).xml')))">
```

ViktorHofer replied:
Please avoid these duplicated NormalizePath msbuild functions as they aren't cost free. You could instead define another two properties next to `XmlDocFileRoot´ for that purpose:

```xml
<XmlDocFileFromNetFolder>$([MSBuild]::NormalizePath('$(XmlDocFileRoot)', 'net', '1033', '$(AssemblyName).xml')</XmlDocFileFromNetFolder>
<XmlDocFileFromDotNetPlatExtFolder>$([MSBuild]::NormalizePath('$(XmlDocFileRoot)', 'dotnet-plat-ext', '1033', '$(AssemblyName).xml')</XmlDocFileFromDotNetPlatExtFolder>
```

ViktorHofer commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,41 @@
+  </PropertyGroup>
+
+  <!-- If the project is not generating its own documentation yet, retrieve it from the dotnet-api-docs nuget package. -->
+  <Choose>
+    <When Condition="('$(UseIntellisenseDocumentationFile)' == '' or '$(UseIntellisenseDocumentationFile)' == 'false') and Exists($([MSBuild]::NormalizePath('$(XmlDocFileRoot)', 'net', '1033', '$(AssemblyName).xml')))">
```
Unfortunately "Choose / When" constructs are quite verbose in msbuild syntax which is why they are hardly used.
You don't gain anything by using them here as you can achieve the same with a far less verbose syntax. Put the `"'$(UseIntellisenseDocumentationFile)' != 'true'"` condition on the property group and then use the `Exists` function on the two property conditions inside it.

ericstj replied:
IMO we don't even need `UseIntellisenseDocumentationFile` to be considered in these properties.  Setting the value for IntellisenseFilePath can be done unconditionally.  We might even just move it inside ChangeDocumentationFile... target since it's only ever used there.

ViktorHofer replied:
> IMO we don't even need UseIntellisenseDocumentationFile to be considered in these properties. Setting the value for IntellisenseFilePath can be done unconditionally

We only want to use the Exists msbuild function when necessary as it results in an underlying IO call. Setting these properties unconditionally means that we always have at least one IO call vs. when conditioning on `UseIntellisenseDocumentationFile`, we would avoid that when the intellisense package file isn't used.

> We might even just move it inside ChangeDocumentationFile... target since it's only ever used there.

... which makes the Exists msbuild function conditional and that's what we want. Either inside or outside the target works for me. It boils down to if we want to do the Exists check lazily or during evaluation time. Lazily is probably better to avoid unnecessary work during design time and restore builds.

carlossanlop committed:
Move PackageDownload to intellisense.targets

carlossanlop committed:
Remove Choose+When. Readjust properties for Private.Intellisense files/folders. Fix conditions for doc file swap.

carlossanlop committed:
Rename csproj property to something more clear.

carlossanlop committed:
Default csproj property to true.

carlossanlop committed:
Missed adding the "not previously set" condition for UseIntellisensePackageDocXmlFile in intellisense.targets.

carlossanlop commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,37 @@
+
+  <Target Name="CopyDocumentationFileToXmlDocDir"
+          AfterTargets="CopyFilesToOutputDirectory"
+          Condition="'$(UseIntellisensePackageDocXmlFile)' == 'true' and '$(IsNetCoreAppSrc)' == 'true' and '$(TargetFramework)' == '$(NetCoreAppCurrent)'"
```
I thought this condition was missing a check for UseIntellisensePackageDocXmlFIle, but it's wrong. Need to remove it, we _always_ want to copy _a_ file.

carlossanlop committed:
Remove incorrect condition in CopyDocumentationFileToXmlDocDir.

carlossanlop commented:
#### System.Formats.Cbor (triple slash + OOB)
- Pass: `artifacts\obj\System.Formats.Cbor\Debug\net8.0` - Local xml.
- Pass: `artifacts\bin\System.Formats.Cbor\Debug\net8.0` - Local xml.
- Pass: `dotnet pack: artifacts\packages\Debug\Shipping\System.Formats.Cbor.8.0.0-dev.nupkg\lib\net8.0` - Local xml.
- Fail: `artifacts\bin\docs` - No file copied.

#### System.Numerics.Vectors (triple slash)
- Fail: `artifacts\obj\System.Numerics.Vectors\Debug\net8.0` - Empty xml (no APIs).
- Fail: `artifacts\bin\System.Numerics.Vectors\Debug\net8.0` - Empty xml (no APIs).
- Skip: `dotnet pack`
- Fail: `artifacts\bin\docs` - Empty xml (no APIs).

#### System.Formats.Tar (docs package)
- Fail: `artifacts\obj\System.Formats.Tar\Debug\net8.0` - Xml with only SR strings.
- Fail: `artifacts\bin\System.Formats.Tar\Debug\net8.0` - Xml with only SR strings.
- Skip: `dotnet pack`
- Pass: `artifacts\bin\docs` - Docs package xml.

#### System.IO.Compression.Brotli (triple slash)
- Fail: `artifacts\obj\System.IO.Compression.Brotli\Debug\net8.0` - Xml with only SR strings.
- Fail: `artifacts\bin\System.IO.Compression.Brotli\Debug\net8.0` - Xml with only SR strings.
- Skip: `dotnet pack`
- Fail: `artifacts\bin\docs` - Xml with only SR strings.

ViktorHofer commented:
> I noticed the property IsNetCoreAppSrc is not available in intellisense.targets when I build an individual csproj with dotnet build. Is that expected? It's preventing me from getting successful results.

Yes that's expected as you import the intellisense.targets file too early. Consider moving it somewhere [here](https://github.com/dotnet/runtime/blob/7cc9aedcce35758e94cae6e0aa9af562c8bcf800/src/libraries/Directory.Build.targets#L79) to have access to that property.

ViktorHofer commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,37 @@
+    <IntellisensePackageXmlFileFromNetFolder>$([MSBuild]::NormalizePath('$(IntellisensePackageXmlRootFolder)', 'net', '1033', '$(AssemblyName).xml'))</IntellisensePackageXmlFileFromNetFolder>
+    <IntellisensePackageXmlFileFromDotNetPlatExtFolder>$([MSBuild]::NormalizePath('$(IntellisensePackageXmlRootFolder)', 'dotnet-plat-ext', '1033', '$(AssemblyName).xml'))</IntellisensePackageXmlFileFromDotNetPlatExtFolder>
+    <IntellisensePackageXmlFile Condition="'$(UseIntellisensePackageDocXmlFile)' == 'true' and Exists($(IntellisensePackageXmlFileFromNetFolder))">$(IntellisensePackageXmlFileFromNetFolder)</IntellisensePackageXmlFile>
+    <IntellisensePackageXmlFile Condition="'$(UseIntellisensePackageDocXmlFile)' == 'true' and Exists($(IntellisensePackageXmlFileFromDotNetPlatExtFolder))">$(IntellisensePackageXmlFileFromDotNetPlatExtFolder)</IntellisensePackageXmlFile>
```
For cases where an xml file exists in both folders (net & dotnet-plat-ext), you want to pick it from the former location.

```suggestion
    <IntellisensePackageXmlFile Condition="'$(IntellisensePackageXmlFile)' == '' and '$(UseIntellisensePackageDocXmlFile)' == 'true' and Exists($(IntellisensePackageXmlFileFromDotNetPlatExtFolder))">$(IntellisensePackageXmlFileFromDotNetPlatExtFolder)</IntellisensePackageXmlFile>
```

carlossanlop replied:
Good point.

Took me a minute to understand your suggestion, but I see it now: if the first conditioned property did not acquire a value, evaluate the second.

ViktorHofer commented:
Apart from the remaining comments, this looks good. Ideally, this PR would also enable `CS1591` warnings. Now that the Arcade change is already consumed, that is easily done via the following steps:
1. Set `<SkipArcadeNoWarnCS1591>true</SkipArcadeNoWarnCS1591>` in eng/Versions.props, near the other features switches at the top.
2. Add a `<NoWarn Condition="'$(UseIntellisensePackageDocXmlFile)' == 'true'">$(NoWarn);1591</NoWarn>` property to intellisense.targets, somewhere below `<UseIntellisensePackageDocXmlFile>`.

With that, you should now see CS1591 errors in the projects that set `UseIntellisensePackageDocXmlFile` if public API isn't annotated.

carlossanlop committed:
Missed property reuse in Condition in sfxproj

carlossanlop committed:
Only evaluate second IntellisensePackageXmlFile if the first one did not acquire any value.

carlossanlop committed:
Move intellisense.targets import from root to libraries, right after the IsNetCoreAppSrc property is declared so we can use it.

carlossanlop commented:
I tested all 4 assemblies as in my [previous comment](https://github.com/dotnet/runtime/pull/79134#issuecomment-1405940505) with the newest commits. I got the exact same results.

carlossanlop committed:
Enable CS1591, skip arcade warning

carlossanlop commented:
Doing a full build after enabling CS1591 shows that System.Private.CoreLib is not respecting the condition to check for UseIntellisensePackageDocXmlFile and disable CS1591.

carlossanlop committed:
Move SkipArcadeNoWarnCS1591 to src/libraries/Directory.Build.props to avoid coreclr failures

ViktorHofer commented:
> So I decided to move SkipArcadeNoWarnCS1591 from eng/Versions.props to src/libraries/Directory.Build.props, to ensure we only enable the rule in libraries, and to avoid duplication of properties.

Unfortunately that won't work as Arcade needs it set before the Arcade.Sdk is imported. If you want to keep it in src/libraries/Directory.Build.props, you will need to move it somewhere before L7 (where the `<Import Project="..\..\Directory.Build.props" />` happens.

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Ref.sfxproj:
```diff
@@ -14,7 +14,7 @@
     <ItemGroup>
       <ReferencePath Include="@(LibrariesRefAssemblies)" />
-      <DocFilesToPackage Include="$(ArtifactsBinDir)/docs/%(LibrariesRefAssemblies.FileName).xml" Condition="Exists('$(ArtifactsBinDir)/docs/%(LibrariesRefAssemblies.FileName).xml')"/>
+      <DocFilesToPackage Include="$(XmlDocDir)/%(LibrariesRefAssemblies.FileName).xml" Condition="Exists('$(XmlDocDir)/%(LibrariesRefAssemblies.FileName).xml')"/>
```
nit: you don't the slash here as `XmlDocDir` already has a trailing slash, but we can keep it as msbuild will just ignore it.

carlossanlop commented:
@ericstj Viktor and I have a question: How should we handle typeforwards/partial facade assemblies? The docs files are showing up as empty (no APIs) for System.Numerics.Vectors when I do `dotnet build src/System.Numerics.Vectors.csproj -f net8.0`.

carlossanlop committed:
Remove unnecessary slashes in sfxproj

carlossanlop committed:
Move SkipArcadeNoWarn up to the top of src/libraries/Directory.Build.props to ensure it gets loaded before importing arcade.

carlossanlop commented:
Pending trying to find out why Brotli is only generating SR files (Viktor is helping me).

ericstj commented:
The compiler isn't going to emit those for you, you'll need to merge them from the dependent assembly.  Maybe through a build task or source generator or something?  @safern might remember some thinking around that.

ViktorHofer commented:
> Pending trying to find out why Brotli is only generating SR files (Viktor is helping me).

Broti only has SR resources because the net8.0 tfm is a PNSE assembly: https://github.com/dotnet/runtime/blob/7d4163fc6e4b5c634eef22d0fd8decfc2aaca763/src/libraries/System.IO.Compression.Brotli/src/System.IO.Compression.Brotli.csproj#L9

That means that it will automatically generate public API with throwing body based on the contract/reference assembly's source. The reference assembly's source isn't XML doc annotated (by design). We will probably need a design discussion first how to handle this. In https://github.com/dotnet/runtime/issues/44969, Santi talks about that problem a bit.

IMHO ideally (because that would also solve some other scenarios like the ref <--> src project convergence) we would stop generating PNSE source code automatically and use checked-in code instead. I would imagine that we could either
- migrate existing PNSE auto-generated code into runtime condition (i.e. `if (OperatingSystem.IsWindows) ..., else if (...), else { throw PNSE(); }`
- or if we want to keep building for different platforms, use compiler preprocessor directives:
```c#
/// <summary>
/// This is a public api method
/// </summary>
public void PublicApiMethod()
{
#if PNSE
    throw new PlatfornNotSupportedException();
#else
    PublicApiMethodImpl();   
#endif
}

These are just some random thoughts from my side. I strongly support any design that moves away from auto-generated PNSE code as that will also solve other hard problems.

ViktorHofer commented:
from @ericstj:
> The compiler isn't going to emit those for you, you'll need to merge them from the dependent assembly. Maybe through a build task or source generator or something? @safern might remember some thinking around that.

It would be great if we could find a solution that would contribute to https://github.com/dotnet/runtime/issues/58163 as well:

from https://github.com/dotnet/runtime/issues/58163:
> It's yet unclear how to generate these type forwards without having a reference assembly available that represents the public API surface area. A possible solution to that could be to check in the type forwards, adding them to the compilation input and making sure that they are kept up to date. The problem with that is that Roslyn keeps passed-in type forward attributes in reference assemblies instead of following the type and replacing the attribute with the actual type. We should sync up with the Roslyn team to see if supporting this scenario is feasible.

Overall, I see https://github.com/dotnet/runtime/issues/58163 and this effort tangled. By finding the right design, we should be able to contribute to both efforts at the same time.

safern commented:
Joining the party a little late but on Friday I did not get a chance to reply.

When we were looking at this a couple of years ago and based on the discussion above, I think there are 3 main scenarios to consider for generating the XML files:

1. PNSE assemblies which use the `GeneratePlatformNotSupportedMessage` feature.
2. Facade assemblies which have at least one typeforward.
3. Cross RID assemblies which have different runtime implementations.

I think we only had an idea on how to solve for 2 and 3.

For facade assemblies and Cross RID assemblies we said that we were going to add an MSBuild Task that would run ONLY on the outer build to not affect the vertical build times and also we need the outputs of all the OSs. The MSBuild Task would do something like:

1. For facade assemblies:
     - get all the typeforwards generated (maybe choose the built asset for the matching TargetOS or the Host OS, since we are in an outerbuild), get the generated XML file with the docs for all it's references and get the matching docs for the types that were forwarded and move them over to the facade assembly's main doc xml.
     - I think this could also be done in a source generator, but that would run for all inner builds, but might be "simpler" implementation wise?
     - I don't know if this could be a `Roslyn` feature that could be built via an opt-in switch, something like, `GenerateXmlDocsForTypeForward` and also how feasible since `Roslyn` would need to resolve the docs for the forwarded types.

2. For cross RID assemblies we agreed that we were going to turn on the doc generation and also the warning that all public API must have a triple slash document so that we ensure all the OSs implementations had triple slash comments, however we need to include some validation to make sure we produce the same XML output regardless of the OS it is being built on, or this could be optimized to only be done when in a local developer build or an all configuration build to not affect the CI verticals. The task would do something like:
    - Validate all xml files for all OS builds are the same. I don't know if this could just do a `git diff` (that would require all contributors to have `git` installed on the path, but would perhaps be simpler and more efficiente than doing it on an MSBuild Task? It could also be text diff if that meets the criteria.
    - Choose any produced xml file (perhaps the one applicable to the Target OS/Host OS) as we already ensured that all of them are the same.

For the `PNSE` scenario I see 2 options:
1. Check-in the XML file with all the documentation for the APIs and perhaps add validation that all the public APIs in the assembly are present in the XML file.
2. As Viktor mentioned above, remove the magic PNSE generator (that would potentially be needed if we ever wanted to get RID of the ref assemblies' source code) and check-in the source code that throws the PNSE explicitly, if it does for all OSs then a simple PNSE source file, if not, it could be included via compile conditions on OS/TFM and then the checked-in source code would need to have the triple slash comments. 

But I think these are incremental steps that could be done separately, and you could potentially start just having the source of truth being triple slash for assemblies that don't use `IsPartialFacadeAssembly` and don't use `GeneratePlatformNotSupportedMessage`.

carlossanlop commented on src/libraries/System.Formats.Cbor/src/System.Formats.Cbor.csproj:
```diff
@@ -3,6 +3,7 @@
     <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <IsPackable>true</IsPackable>
+    <UseIntellisensePackageDocXmlFile>false</UseIntellisensePackageDocXmlFile>
```
Need to remove this property enabling the feature in the 3 assemblies, for now.

ViktorHofer replied:
Cbor doesn't use `IsPartialFacadeAssembly`, nor does it use `GeneratePlatformNotSupportedAssemblyMessage` so that property can stay here and we can use the project's produced intellisense file.

carlossanlop replied:
Ah right, it was the only assembly that passed the test results.

carlossanlop replied:
Alrught, I will enable it for the Brotli assembly but in a follow-up PR because it is currently failing due to some public APIs not being documented. I don't want to address all those errors here.

ViktorHofer commented:
@carlossanlop as we discussed offline, please add a validation target in intellisense.targets that emits an error if `UseIntellisensePackageDocXmlFile` is false (means that the project opts into using the compiler produced xml file) and the project either sets `IsPartialFacadeAssembly` to true or when `GeneratePlatformNotSupportedAssemblyMessage` isn't empty.

carlossanlop committed:
Revert Brotli and Vectors csproj changes. Those assemblies are either a partial facade or use PNSE.

carlossanlop committed:
Add target to intellisense.targets that runs as InitialTarget of the current project, and throws errors if the UseIntellisensePackageDocXmlFile is set to false and the assembly is either a partial facade or uses PNSE.

carlossanlop commented:
Below is a test of the latest 2 commits.

The error message shows up in assemblies where I temporarily set `UseIntellisensePackageDocXmlFile` to `false`, and they are either partial facades or use PNSE. The only one that does not have neither of those conditions is Cbor. It's the first result, built successfully, and is the only one we will merge with `UseIntellisensePackageDocXmlFile` set to `false` with this PR.

```
PS D:\runtime\src\libraries> dotnet build .\System.Formats.Cbor\src\System.Formats.Cbor.csproj -f net8.0 /bl 
  MSBuild version 17.5.0-preview-23067-01+771199907 for .NET
    Determining projects to restore...
    All projects are up-to-date for restore.
    Microsoft.Interop.SourceGeneration -> D:\runtime\artifacts\bin\Microsoft.Interop.SourceGeneration\Debug\netstandard2.0\Microsoft.Interop.SourceGeneration.dll
    LibraryImportGenerator -> D:\runtime\artifacts\bin\LibraryImportGenerator\Debug\netstandard2.0\Microsoft.Interop.LibraryImportGenerator.dll
    System.Formats.Cbor -> D:\runtime\artifacts\bin\System.Formats.Cbor\ref\Debug\net8.0\System.Formats.Cbor.dll
    System.Formats.Cbor -> D:\runtime\artifacts\bin\System.Formats.Cbor\Debug\net8.0\System.Formats.Cbor.dll

  Build succeeded.
      0 Warning(s)
      0 Error(s)

  Time Elapsed 00:00:01.31

PS D:\runtime\src\libraries> dotnet build .\System.Numerics.Vectors\src\System.Numerics.Vectors.csproj -f net8.0 /bl          
  MSBuild version 17.5.0-preview-23067-01+771199907 for .NET
  D:\runtime\eng\intellisense.targets(14,5): error : Cannot currently generate full xml documentation for an assembly that is a partial facade. [D:\runtime\src\libraries\System.Numeric
  s.Vectors\src\System.Numerics.Vectors.csproj]

PS D:\runtime\src\libraries> dotnet build .\System.IO\src\System.IO.csproj -f net8.0 /bl          
  MSBuild version 17.5.0-preview-23067-01+771199907 for .NET
  D:\runtime\eng\intellisense.targets(14,5): error : Cannot currently generate full xml documentation for an assembly that is a partial facade. [D:\runtime\src\libraries\System.IO\src\
  System.IO.csproj]

PS D:\runtime\src\libraries> dotnet build .\System.Formats.Tar\src\System.Formats.Tar.csproj -f net8.0 /bl  
  MSBuild version 17.5.0-preview-23067-01+771199907 for .NET
  D:\runtime\eng\intellisense.targets(15,5): error : Cannot currently generate full xml documentation for an assembly that generates PNSE. [D:\runtime\src\libraries\System.Formats.Tar\
  src\System.Formats.Tar.csproj]

PS D:\runtime\src\libraries> dotnet build .\System.IO.Compression.Brotli\src\System.IO.Compression.Brotli.csproj -f net8.0 /bl
  MSBuild version 17.5.0-preview-23067-01+771199907 for .NET
  D:\runtime\eng\intellisense.targets(15,5): error : Cannot currently generate full xml documentation for an assembly that generates PNSE. [D:\runtime\src\libraries\System.IO.Compressi
  on.Brotli\src\System.IO.Compression.Brotli.csproj]

PS D:\runtime\src\libraries> dotnet build .\System.IO.Ports\src\System.IO.Ports.csproj -f net8.0 /bl                
  MSBuild version 17.5.0-preview-23067-01+771199907 for .NET
  D:\runtime\eng\intellisense.targets(15,5): error : Cannot currently generate full xml documentation for an assembly that generates PNSE. [D:\runtime\src\libraries\System.IO.Ports\src
\System.IO.Ports.csproj]

PS D:\runtime\src\libraries> dotnet build .\System.Drawing.Common\src\System.Drawing.Common.csproj -f net8.0 /bl
  MSBuild version 17.5.0-preview-23067-01+771199907 for .NET
  D:\runtime\eng\intellisense.targets(14,5): error : Cannot currently generate full xml documentation for an assembly that is a partial facade. [D:\runtime\src\libraries\System.Drawing
  .Common\src\System.Drawing.Common.csproj]
```

smasher164 commented on eng/intellisense.targets:
```diff
@@ -10,6 +10,11 @@
   </PropertyGroup>
 
+  <Target Name="VerifyAssemblySupportsDocsXmlGeneration">
+    <Error Condition="'$(UseIntellisensePackageDocXmlFile)' == 'false' and '$(IsPartialFacadeAssembly)' == 'true'" Text="Cannot currently generate full xml documentation for an assembly that is a partial facade." />
```
It might be worth it to mention the assembly name for which we can't generate documentation.

smasher164 commented on eng/intellisense.targets:
```diff
@@ -10,6 +10,11 @@
 
+  <Target Name="VerifyAssemblySupportsDocsXmlGeneration">
+    <Error Condition="'$(UseIntellisensePackageDocXmlFile)' == 'false' and '$(IsPartialFacadeAssembly)' == 'true'" Text="Cannot currently generate full xml documentation for an assembly that is a partial facade." />
+    <Error Condition="'$(UseIntellisensePackageDocXmlFile)' == 'false' and '$(GeneratePlatformNotSupportedAssemblyMessage)' != ''" Text="Cannot currently generate full xml documentation for an assembly that generates PNSE." />
```
Instead of saying `generates PNSE`, we should say `throws PlatformNotSupportedException`.

smasher164 commented on eng/intellisense.targets:
```diff
@@ -10,6 +10,11 @@
   </PropertyGroup>
 
+  <Target Name="VerifyAssemblySupportsDocsXmlGeneration">
+    <Error Condition="'$(UseIntellisensePackageDocXmlFile)' == 'false' and '$(IsPartialFacadeAssembly)' == 'true'" Text="Cannot currently generate full xml documentation for an assembly that is a partial facade." />
```
s/xml/XML/g
Also, I don't think we need "currently".

carlossanlop replied:
Message reworded.

carlossanlop committed:
Improve error message: Mention the problematic property and the offending assembly.

ViktorHofer commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,47 @@
+    <IntellisensePackageXmlFile Condition="'$(IntellisensePackageXmlFile)' == '' and '$(UseIntellisensePackageDocXmlFile)' == 'true' and Exists($(IntellisensePackageXmlFileFromDotNetPlatExtFolder))">$(IntellisensePackageXmlFileFromDotNetPlatExtFolder)</IntellisensePackageXmlFile>
+  </PropertyGroup>
+
+  <Target Name="VerifyAssemblySupportsDocsXmlGeneration">
```
InitalTargets are expensive as they run for all project evaluations. Therefore, please add a target condition (i.e. to only run when the UseIntellisense... prop is false).

carlossanlop replied:
Ok, good to know.

carlossanlop replied:
I made the change. So if I understood this correctly, I can keep "InitialTargets" as long as the target is conditioned, which is what you requested to add.

Tested manually with and without the UseIntellisense... property set to false in various csproj files, and the results were the expected ones.

carlossanlop committed:
Added extra condition to verification target to only run when UseIntellisense... is not set to true.
Also moved the initial definition to its own solitary property group right before the verification target runs.

Rename the properties that acquire the value of a file path to "FilePath" instead of just "File", for clarity.

carlossanlop commented:
CI failure is unrelated and pre-existing: https://github.com/dotnet/runtime/issues/48798
The manual tests in various csproj with and without enabling the new property, worked as expected.
Signed-off, so I'm merging now. :shipit:

ViktorHofer commented on eng/intellisense.targets:
```diff
@@ -0,0 +1,51 @@
+
+  <Target Name="CopyDocumentationFileToXmlDocDir"
+          AfterTargets="CopyFilesToOutputDirectory"
+          Condition="'$(IsNetCoreAppSrc)' == 'true' and '$(TargetFramework)' == '$(NetCoreAppCurrent)'"
```
Just realized that the `'$(TargetFramework)' == '$(NetCoreAppCurrent)'` condition results in a behavior difference and will force us to overbuild. When building the repo, by default only the nearest NetCoreAppCurrent tfm will be built. In the case of building on Windows, this means net8.0-windows or net8.0 if projects don't have a corresponding windows platform TFM.

We should remove that artificial limitation and restore the previous behavior. I could see the following cases to apply:
- A project doesn't multi-target => use NetCoreAppCurrent condition (status quo).
- A project multi-targets and all inner builds are built (no tfm filtering) => run this target in the NetCoreAppCurrent inner build (status quo).
- A project multi-targets but an inner build runs stand-alone (tfm filtering) => Don't condition on a TFM.

I don't know if it's possible to determine the third case in an msbuild environment, but that's the default build behavior in dotnet/runtime and will be utilized in the VMR with RID builds. Therefore we should try to make that work.

carlossanlop replied:
I created an issue to track this: https://github.com/dotnet/runtime/issues/82191
